miden-lib/asm/sat/epilogue.masm

```diff
@@ -0,0 +1,229 @@
+
+# BOOKKEEPING
+# -------------------------------------------------------------------------------------------------
+
+# The memory address at which the number of created notes is stored.
+const.NUM_CREATED_NOTES_PTR=2
+
+# GLOBAL MEMORY
+# -------------------------------------------------------------------------------------------------
+
+# The memory address at which the initial nonce is stored.
+const.INIT_NONCE=14
+
+# CREATED NOTES
+# -------------------------------------------------------------------------------------------------
+
+# The memory address at which the created notes section begins.
+const.CREATED_NOTE_SECTION_OFFSET=10000
+
+# The offsets at which data of a created note is stored relative to the start of its data segment.
+const.CREATED_NOTE_META_OFFSET=0
+const.CREATED_NOTE_RECIPIENT_OFFSET=1
+const.CREATED_NOTE_VAULT_HASH_OFFSET=2
+const.CREATED_NOTE_NUM_ASSETS_OFFSET=3
+const.CREATED_NOTE_ASSETS_OFFSET=4
+
+# CREATED NOTES PROCEDURES
+# =================================================================================================
+
+# TODO: Consider changing the crated notes procedures to work with pointers. Instead of taking an
+# index as input, we could take a pointer to the created note data. This would allow us to avoid
+# the need to compute the pointer multiple times across the different procedures.
+
+#! Computes the vault hash of the created note with index i. The hash is computed as a sequential
+#! hash of the assets contained in the note. If there are an odd number of assets, then for the
+#! final hashing permutation we pad the last word of the hasher rate with ZERO's.
+#!
+#! Stack: [i]
+#! Output: [VAULT_HASH]
+#!
+#! - i is the index of the created note.
+#! - VAULT_HASH is the vault hash of the created note with index i.
+proc.compute_created_note_vault_hash
+    # get created note data pointer from created note index
+    exec.utils::created_note_data_ptr
+
+    # duplicate created note pointer - we will use this to save vault hash to memory
+    dup
+
+    # get the number of assets from memory
+    dup push.CREATED_NOTE_NUM_ASSETS_OFFSET add mem_load
+
+    # round up the number of assets to the next multiple of 2 (simplifies reading of assets)
+    dup is_odd
+
+    # pad to multiple of two if odd
+    if.true
+        push.1 add
+    end
+
+    # initiate counter for assets
+    push.0
+
+    # prepare address and stack for reading assets
+    movup.2 push.CREATED_NOTE_ASSETS_OFFSET add padw padw padw
+
+    # check if the number of assets is greater then 0
+    dup.14 dup.14 neq
+
+    # loop and read assets from memory
+    while.true
+        # read assets from memory.
+        # if this is the last permutation of the loop and we have an odd number of assets then we
+        # implicitly pad the last word of the hasher rate with ZERO's by reading from empty memory.
+        mem_stream
+
+        # check if we should loop again
+        movup.13 push.2 add dup movdn.14 dup.15 neq
```

Comment:
Comment for the future: since this is the place where we iterate over all assets of created notes, this might be the place where we construct the final transaction vault. This will require adjusting semantics of `mem_stream` instruction somewhat (so that it doesn't load and hash, but only does load).

Should we leave a TODO here for this?